bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryDirectDataWriterHelper.java

```diff
@@ -184,19 +194,81 @@ public void addRow(ByteString message) throws IOException {
     appendRequestRowCount++;
   }
 
+  /**
+   * Thorws an exception if there is an error in any of the responses received thus far. Since
+   * responses arrive in order, we proceed to check the next response only after the previous
+   * response has arrived.
+   *
+   * @throws BigQueryConnectorException If there was an invalid response
+   */
+  private void checkForFailedResponse(boolean waitForResponse) {
+    while (nextIndexCheckResponse < appendRowsFutures.size()) {
+      ApiFuture<AppendRowsResponse> validatedAppendRowsFuture =
+          appendRowsFutures.get(nextIndexCheckResponse);
+      if (waitForResponse || validatedAppendRowsFuture.isDone()) {
+        boolean succeeded = false;
+        try {
+          AppendRowsResponse response = validatedAppendRowsFuture.get();
+          // If we got here, the response was successful
+          succeeded = true;
+          nextIndexCheckResponse++;
+        } catch (ExecutionException e) {
+          if (e.getCause().getClass() == OffsetAlreadyExists.class) {
+            // Under heavy load the WriteAPI client library may resend a request due to failed
+            // connection. The OffsetAlreadyExists exception indicates a request was previously
+            // sent, hence it is safe to ignore it.
+            logger.warn("Ignoring OffsetAlreadyExists exception: {}", e.getCause().getMessage());
+            succeeded = true;
+            nextIndexCheckResponse++;
+          } else {
+            logger.error(
+                "Write-stream {} with name {} exception while retrieving AppendRowsResponse",
+                partitionId,
+                writeStreamName);
+            throw new BigQueryConnectorException(
+                "Exception while retrieving AppendRowsResponse", e);
+          }
+        } catch (InterruptedException e) {
+          logger.error(
+              "Write-stream {} with name {} exception while retrieving AppendRowsResponse",
+              partitionId,
+              writeStreamName);
+          throw new BigQueryConnectorException("Exception while retrieving AppendRowsResponse", e);
+        } finally {
+          if (!succeeded) {
+            appendRowsExecutor.shutdown();
+          }
+        }
+      } else {
+        break;
+      }
+    }
+  }
+
   /**
    * Sends an AppendRowsRequest to the BigQuery Storage Write API.
    *
    * @throws IOException If the append rows request fails: either by returning the wrong offset
    *     (deduplication error) or if the response contains an error.
    */
   private void sendAppendRowsRequest() throws IOException {
+    boolean waitForResponse = false;
+    if (writeAtLeastOnce) {
+      // FUTURE: Enable async writes with default stream once client retries are supported. Without
+      // client retries we can encounter vortex stream errors.
```

Comment:
What are vortex stream errors, and how is writing exactly once immune to them?

Reply:
Vortex is the code name for the server engine that ingests data. I have revised the comment to remove the reference to this name. Default stream is more resource-efficient because it enables multiple connections to share the same backend ingestion stream. In contrast, for a non-default stream there is one ingestion stream per connection. When sharing, the backend automatically adjusts the level of sharing in response to traffic volume. As it transitions there may be re-tryable errors returned.